### PR TITLE
Use aid.txt for getting analysis id

### PR DESCRIPTION
**Pull request by atombrella**

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/codesonar/CodeSonarPublisher.java [hunk 1]
```diff
@@ -135,14 +160,16 @@ public void perform(@Nonnull Run<?, ?> run, @Nonnull FilePath workspace, @Nonnul
...
-        if (analysisUrl == null) {
-            analysisUrl = analysisService.getLatestAnalysisUrlForAProject(baseHubUri, expandedProjectName);
+        String aid = this.getAid(run, workspace, expandedProjectName);
+        if(aid == null){
+            // Can't get the analysis ID, so let's mark the build as aborted
+            run.setResult(hudson.model.Result.ABORTED);
         }
-        // here we also need to change something
+
         Analysis analysisWarnings = analysisService.getAnalysisFromUrlWarningsByFilter(analysisUrl);
```
Comment (gt-dbechtold) [inline src/main/java/org/jenkinsci/plugins/codesonar/CodeSonarPublisher.java:169]: Shouldn't `analysisUrl` be generated from `aid` and not from the log?  Shouldn't we update `analysisUrl` here based on `aid`?

File: src/main/java/org/jenkinsci/plugins/codesonar/CodeSonarPublisher.java [hunk 2]
```diff
@@ -98,6 +98,31 @@ public String getVisibilityFilter() {
         return this.visibilityFilter;
     }
 
+    // projectName may be a
```
Comment (gt-dbechtold) [inline src/main/java/org/jenkinsci/plugins/codesonar/CodeSonarPublisher.java:101]: Not sure what this comment is intended to express:

    // projectName may be a

File: src/main/java/org/jenkinsci/plugins/codesonar/CodeSonarPublisher.java [hunk 3]
```diff
@@ -98,6 +98,31 @@ public String getVisibilityFilter() {
...
+    private String getAid(@Nonnull Run<?,?> run, @Nonnull FilePath workspace, @Nonnull String projectName) throws java.io.IOException {
```
Comment (gt-dbechtold) [inline src/main/java/org/jenkinsci/plugins/codesonar/CodeSonarPublisher.java:102]: Can we use an optional, `@DataBoundSetter` property to allow the user to specify a CodeSonar proj_files directory?  That way, we wouldn't need to assume its location.

It might be easier to provide an optional property to specify the AID.  That would put the burden on the user to find the AID themselves (perhaps by looking in the prj_files directory themselves), but it seems like the plugin code would be less fragile if we did it this way.

We could also provide both options - AID or prj_files directory path, but if both were specified at once, you'd need to raise an exception.

Basically, I think adding an AID option is low risk, but it is more difficult for the end-user.  Adding a prj_files directory option would be easier for the end-user, but it would be a higher risk feature.  I think if you go through the trouble of writing a prj_files option, then an AID option would be very easy and it might be a good fallback in case there are problems with the prj_files option.

File: src/main/java/org/jenkinsci/plugins/codesonar/CodeSonarPublisher.java (hunk 2 above)
Comment (atombrella) [reply src/main/java/org/jenkinsci/plugins/codesonar/CodeSonarPublisher.java:101]: Thanks. I'll remove it.

File: src/main/java/org/jenkinsci/plugins/codesonar/CodeSonarPublisher.java (hunk 3 above)
Comment (atombrella) [reply src/main/java/org/jenkinsci/plugins/codesonar/CodeSonarPublisher.java:102]: I can do that, but I fear it might cause some confusion. In the manual, under **Build/Analysis Commands**, there are some pointers that I should remember to include, e.g., the path/to parameter for the analyze command.

Most likely, the chosen default in the function is fine in most cases, I think.

File: src/main/java/org/jenkinsci/plugins/codesonar/CodeSonarPublisher.java [hunk 4]
```diff
@@ -98,6 +98,31 @@ public String getVisibilityFilter() {
...
+        /**
+         * The analysis ID from the aid.txt file
+         */
+        String aid = workspace.asCallableWith(new hudson.FilePath.FileCallable<String>() {
+            @Override
+            public void checkRoles(org.jenkinsci.remoting.RoleChecker roleChecker) throws SecurityException {
```
Comment (atombrella) [inline src/main/java/org/jenkinsci/plugins/codesonar/CodeSonarPublisher.java:108]: Just as a note. We don't check access to workspace here. Probably should be done.

**Comment by jprice-gt** [comment]
If this is going to be the way forward, needing to have the .prj_files directory in the workspace, then the documentation will need to be updated to state that this is a requirement.
